finetune_fsdp.py

```diff
@@ -0,0 +1,849 @@
+                else:
+                    non_quantized_state_dict[name] = tensor
+            assert len(all_quantized_weight_names) == 0, f"mismatched names: {all_quantized_weight_names}"
+            torch.save(non_quantized_state_dict, os.path.join(output_path, "non_quantized_state_dict.pth"))
+            print(f"Saved best model to {output_path}")
+
+
+def main():
+    parser = argparse.ArgumentParser(add_help=True)
+    add_model_args(parser)
+    add_data_args(parser)
+    add_finetuning_args(parser)
+    args = parser.parse_args()
+
+    assert torch.cuda.is_available() and torch.distributed.is_available()
+    torch.distributed.init_process_group()
+    rank = torch.distributed.get_rank()
+    world_size = torch.distributed.get_world_size()
+    device = torch.device(f"cuda:{rank}")
+    torch.cuda.set_device(device)
+
+    assert args.batch_size is not None, "please specify batch size"
+    assert args.batch_size % world_size == 0
+    if args.microbatch_size is None:
+        args.microbatch_size = args.batch_size // world_size
+    assert args.batch_size % (world_size * args.microbatch_size) == 0
+    grad_accumulation_steps = args.batch_size // (world_size * args.microbatch_size)
+    args.load_dtype = getattr(torch, args.load_dtype) if args.load_dtype != 'auto' else 'auto'
+    args.amp_dtype = getattr(torch, args.amp_dtype) if args.amp_dtype is not None else None
+    args.code_dtype = getattr(torch, args.code_dtype) if args.code_dtype is not None else None
+    args.master_dtype = getattr(torch, args.master_dtype)
+    if args.straight_through_buffer_dtype is not None:
+        args.straight_through_buffer_dtype = getattr(torch, args.straight_through_buffer_dtype)
+    else:
+        args.straight_through_buffer_dtype = args.master_dtype
+
+    if args.save_every_steps is not None:
+        assert args.save is not None, f"save_every_steps={args.save_every_steps}, but --save path not specified"
+    if args.keep_best_model:
+        assert args.save is not None, f"--keep_best_model requires --save path"
+        assert args.eval_every_steps is not None, f"--keep_best_model requires --eval_every_steps"
+        assert args.eval_datasets is not None, f"--keep_best_model requires --eval_datasets"
+
+    if args.wandb and rank == 0:
+        assert has_wandb, "`wandb` not installed, try pip install `wandb`"
+        wandb.init(config={a: getattr(args, a) for a in dir(args) if not a.startswith("_")})
+
+    if rank == 0:
+        print(args)
+
+    tokenizer = transformers.AutoTokenizer.from_pretrained(args.base_model)
+    assert tokenizer.eos_token_id is not None
+    tokenizer.pad_token = tokenizer.eos_token
+
+    with master_rank_first(local=True):
+        dataset = prepare_training_dataset(args, tokenizer)
+        if args.save_dataset_and_exit is not None:
+            if rank == 0:
+                dataset.save_to_disk(args.save_dataset_and_exit)
+            exit()
```

Comment:
exit() fails execution, maybe return in all ranks?

Reply:
resolution: will check if this is already done; if not, do